Test failures with Python 3.11

### Description
<!-- Provide a general description of the bug. -->
I'm building astropy 5.1 with a prerelease of python 3.11 for the next Fedora release. I'm getting this error in the tests:

```
=================================== FAILURES ===================================
_______ test_simple_object[0-StaticMatrixTransform-args6-kwargs6-False] ________
pickle_protocol = 0
name = <class 'astropy.coordinates.transformations.StaticMatrixTransform'>
args = [array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]]), <class 'astropy.coordinates.builtin_frames.icrs.ICRS'>, <class 'astropy.coordinates.builtin_frames.icrs.ICRS'>]
kwargs = {}, xfail = False
    @pytest.mark.parametrize(("name", "args", "kwargs", "xfail"),
                             zip(_names, _args, _kwargs, _xfail))
    def test_simple_object(pickle_protocol, name, args, kwargs, xfail):
        # Tests easily instantiated objects
        if xfail:
            pytest.xfail()
        original = name(*args, **kwargs)
>       check_pickling_recovery(original, pickle_protocol)
astropy/coordinates/tests/test_pickle.py:71: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/tests/helper.py:446: in check_pickling_recovery
    generic_recursive_equality_test(original, unpickled,
astropy/tests/helper.py:433: in generic_recursive_equality_test
    generic_recursive_equality_test(dict_a[key],
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
a = <class 'astropy.coordinates.builtin_frames.icrs.ICRS'>
b = <class 'astropy.coordinates.builtin_frames.icrs.ICRS'>
class_history = [<class 'abc.ABCMeta'>, <class 'astropy.coordinates.transformations.StaticMatrixTransform'>]
    def generic_recursive_equality_test(a, b, class_history):
        """
        Check if the attributes of a and b are equal. Then,
        check if the attributes of the attributes are equal.
        """
>       dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
E       TypeError: unbound method object.__getstate__() needs an argument
astropy/tests/helper.py:415: TypeError
```

This is probably related with a change in how objects are pickled. Now `object` has a default implementation of `__getstate__ 

https://docs.python.org/3.11/library/pickle.html#object.__getstate__

Full details in the Fedora bug report: https://bugzilla.redhat.com/show_bug.cgi?id=2101963

--- a/.github/workflows/ci_workflows.yml
+++ b/.github/workflows/ci_workflows.yml
@@ -99,7 +99,7 @@ jobs:
       if: startsWith(matrix.os, 'ubuntu')
       run: |
         sudo apt-get update
-        sudo apt-get install language-pack-fr tzdata
+        sudo apt-get install language-pack-fr tzdata libopenblas-dev
     - name: Install tox
       run: python -m pip install --upgrade tox
     - name: Run tests

--- a/.github/workflows/ci_workflows.yml
+++ b/.github/workflows/ci_workflows.yml
@@ -120,9 +120,9 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - name: (Allowed Failure) Python 3.11-rc1 with remote data and dev version of key dependencies
+          - name: (Allowed Failure) Python 3.11 with remote data and dev version of key dependencies
             os: ubuntu-latest
-            python: '3.11.0-rc.1'
+            python: '3.11.0-rc.2'
             toxenv: py311-test-devdeps
             toxposargs: --remote-data=any
 

--- a/.github/workflows/ci_workflows.yml
+++ b/.github/workflows/ci_workflows.yml
@@ -120,10 +120,10 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - name: (Allowed Failure) Python 3.8 with remote data and dev version of key dependencies
+          - name: (Allowed Failure) Python 3.11-rc1 with remote data and dev version of key dependencies
             os: ubuntu-latest
-            python: 3.8
-            toxenv: py38-test-devdeps
+            python: '3.11.0-rc.1'
+            toxenv: py311-test-devdeps
             toxposargs: --remote-data=any
 
     steps:

--- a/astropy/cosmology/tests/test_core.py
+++ b/astropy/cosmology/tests/test_core.py
@@ -17,10 +17,11 @@ import pytest
 # LOCAL
 import astropy.cosmology.units as cu
 import astropy.units as u
-from astropy.cosmology import Cosmology, CosmologyError, FlatCosmologyMixin
+from astropy.cosmology import Cosmology, FlatCosmologyMixin
 from astropy.cosmology.core import _COSMOLOGY_CLASSES
 from astropy.cosmology.parameter import Parameter
 from astropy.table import Column, QTable, Table
+from astropy.utils.compat import PYTHON_LT_3_11
 from astropy.utils.exceptions import AstropyDeprecationWarning
 from astropy.utils.metadata import MetaData
 
@@ -197,7 +198,9 @@ class TestCosmology(ParameterTestMixin, MetaTestMixin,
         assert cosmo.name == self.cls_kwargs["name"]  # test has expected value
 
         # immutable
-        with pytest.raises(AttributeError, match="can't set"):
+        match = ("can't set" if PYTHON_LT_3_11
+                 else f"property 'name' of {cosmo.__class__.__name__!r} object has no setter")
+        with pytest.raises(AttributeError, match=match):
             cosmo.name = None
 
     def test_is_flat(self, cosmo_cls, cosmo):

--- a/astropy/io/ascii/core.py
+++ b/astropy/io/ascii/core.py
@@ -1088,17 +1088,21 @@ class BaseOutputter:
                 try:
                     col.data = converter_func(col.str_vals)
                     col.type = converter_type
-                except (TypeError, ValueError) as err:
-                    col.converters.pop(0)
-                    last_err = err
-                except OverflowError as err:
+                except (OverflowError, TypeError, ValueError) as err:
                     # Overflow during conversion (most likely an int that
                     # doesn't fit in native C long). Put string at the top of
                     # the converters list for the next while iteration.
-                    warnings.warn(
-                        "OverflowError converting to {} in column {}, reverting to String."
-                        .format(converter_type.__name__, col.name), AstropyWarning)
-                    col.converters.insert(0, convert_numpy(numpy.str))
+                    # With python/cpython#95778 this has been supplemented with a
+                    # "ValueError: Exceeds the limit (4300) for integer string conversion"
+                    # so need to catch that as well.
+                    if isinstance(err, OverflowError) or (isinstance(err, ValueError) and
+                                                          str(err).startswith('Exceeds the limit')):
+                        warnings.warn(
+                            f'OverflowError converting to {converter_type.__name__} in column '
+                            f'{col.name}, reverting to String.', AstropyWarning)
+                        col.converters.insert(0, convert_numpy(numpy.str))
+                    else:
+                        col.converters.pop(0)
                     last_err = err
 
 

--- a/astropy/tests/helper.py
+++ b/astropy/tests/helper.py
@@ -6,6 +6,7 @@ from the installed astropy.  It makes use of the `pytest`_ testing framework.
 import os
 import sys
 import pickle
+import inspect
 import warnings
 import functools
 
@@ -15,7 +16,7 @@ from astropy.units import allclose as quantity_allclose  # noqa: F401
 from astropy.utils.decorators import deprecated
 from astropy.utils.exceptions import (AstropyDeprecationWarning,
                                       AstropyPendingDeprecationWarning)
-
+from astropy.utils.compat import PYTHON_LT_3_11
 
 # For backward-compatibility with affiliated packages
 from .runner import TestRunner  # pylint: disable=W0611  # noqa
@@ -409,11 +410,15 @@ def generic_recursive_equality_test(a, b, class_history):
     Check if the attributes of a and b are equal. Then,
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if PYTHON_LT_3_11:
+        dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    else:
+        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
+        # with required argument(s) are passed to this function.
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__()
     dict_b = b.__dict__
     for key in dict_a:
-        assert key in dict_b,\
-          f"Did not pickle {key}"
+        assert key in dict_b, f"Did not pickle {key}"
 
         if dict_a[key].__class__.__eq__ is not object.__eq__:
             # Only compare if the class defines a proper equality test.

--- a/astropy/time/tests/test_mask.py
+++ b/astropy/time/tests/test_mask.py
@@ -16,7 +16,8 @@ allclose_sec = functools.partial(np.allclose, rtol=2. ** -52,
                                  atol=2. ** -52 * 24 * 3600)  # 20 ps atol
 is_masked = np.ma.is_masked
 
-# This is r"can't set attribute '{0}'" for Python 3.10, but not 3.9.
+# The first form is expanded to r"can't set attribute '{0}'" in Python 3.10, and replaced
+# with the more informative second form as of 3.11 (python/cpython#31311).
 no_setter_err = (r"can't set attribute" if PYTHON_LT_3_11
                  else r"property '{0}' of '{1}' object has no setter")
 

--- a/astropy/time/tests/test_mask.py
+++ b/astropy/time/tests/test_mask.py
@@ -7,6 +7,7 @@ import pytest
 
 from astropy import units as u
 from astropy.utils import iers
+from astropy.utils.compat import PYTHON_LT_3_11
 from astropy.time import Time
 from astropy.table import Table
 from astropy.utils.compat.optional_deps import HAS_H5PY
@@ -14,6 +15,10 @@ from astropy.utils.compat.optional_deps import HAS_H5PY
 allclose_sec = functools.partial(np.allclose, rtol=2. ** -52,
                                  atol=2. ** -52 * 24 * 3600)  # 20 ps atol
 is_masked = np.ma.is_masked
+
+# This is r"can't set attribute '{0}'" for Python 3.10, but not 3.9.
+no_setter_err = (r"can't set attribute" if PYTHON_LT_3_11
+                 else r"property '{0}' of '{1}' object has no setter")
 
 
 def test_simple():
@@ -47,9 +52,8 @@ def test_scalar_init():
 
 def test_mask_not_writeable():
     t = Time('2000:001')
-    with pytest.raises(AttributeError) as err:
+    with pytest.raises(AttributeError, match=no_setter_err.format('mask', t.__class__.__name__)):
         t.mask = True
-    assert "can't set attribute" in str(err.value)
 
     t = Time(['2000:001'])
     with pytest.raises(ValueError) as err:

--- a/astropy/utils/compat/misc.py
+++ b/astropy/utils/compat/misc.py
@@ -11,7 +11,11 @@ import functools
 
 from astropy.utils.decorators import deprecated
 
-__all__ = ['override__dir__', 'possible_filename']
+__all__ = ["override__dir__", "possible_filename",
+           "PYTHON_LT_3_11"]
+
+
+PYTHON_LT_3_11 = sys.version_info < (3, 11)
 
 
 def possible_filename(filename):

--- a/astropy/utils/compat/misc.py
+++ b/astropy/utils/compat/misc.py
@@ -14,7 +14,6 @@ from astropy.utils.decorators import deprecated
 __all__ = ["override__dir__", "possible_filename",
            "PYTHON_LT_3_11"]
 
-
 PYTHON_LT_3_11 = sys.version_info < (3, 11)
 
 

--- a/docs/changes/13614.other.rst
+++ b/docs/changes/13614.other.rst
@@ -1,0 +1,1 @@
+Compatibility with Python 3.11, 3.10.7, 3.9.14, 3.8.14

--- a/tox.ini
+++ b/tox.ini
@@ -28,6 +28,7 @@ setenv =
     !image: MPLFLAGS =
     clocale: LC_CTYPE = C.ascii
     clocale: LC_ALL = C
+    devdeps: PIP_EXTRA_INDEX_URL = https://pypi.anaconda.org/scipy-wheels-nightly/simple
 
 # Run the tests in a temporary directory to make sure that we don't import
 # astropy from the source tree
@@ -87,7 +88,9 @@ deps =
     # The devdeps factor is intended to be used to install the latest developer version
     # or nightly wheel of key dependencies.
     # Do not install asdf-astropy so we can test astropy.io.misc.asdf until we remove it.
-    devdeps,mpldev: git+https://github.com/matplotlib/matplotlib.git#egg=matplotlib
+    devdeps: scipy>=0.0.dev0
+    devdeps: numpy>=0.0.dev0
+    devdeps,mpldev: matplotlib>=0.0.dev0
     devdeps: git+https://github.com/asdf-format/asdf.git#egg=asdf
     devdeps: git+https://github.com/liberfa/pyerfa.git#egg=pyerfa
 
@@ -112,8 +115,6 @@ extras =
     alldeps: test_all
 
 commands =
-    devdeps: pip install -U --pre --only-binary :all: -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy
-    devdeps: pip install -U --pre --only-binary :all: -i https://pypi.anaconda.org/scipy-wheels-nightly/simple scipy
     pip freeze
     !cov-!double: pytest --pyargs astropy {toxinidir}/docs {env:MPLFLAGS} {posargs}
     cov-!double: pytest --pyargs astropy {toxinidir}/docs {env:MPLFLAGS} --cov astropy --cov-config={toxinidir}/pyproject.toml {posargs}

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 envlist =
-    py{38,39,310,dev}-test{,-image,-recdeps,-alldeps,-oldestdeps,-devdeps,-devinfra,-predeps,-numpy118,-numpy119,-numpy120,-numpy121,-mpl311}{,-cov}{,-clocale}
+    py{38,39,310,311,dev}-test{,-image,-recdeps,-alldeps,-oldestdeps,-devdeps,-devinfra,-predeps,-numpy118,-numpy119,-numpy120,-numpy121,-mpl311}{,-cov}{,-clocale}
     build_docs
     linkcheck
     codestyle

